src/combinatorics/hall.lean

```diff
@@ -0,0 +1,341 @@
+    rw card_image_of_injective _ subtype.val_injective },
+  rcases ih r' h3 h4 with ⟨f', hfinj, hfr⟩,
+  refine ⟨λ x, if h : x = a then b else f' ⟨x, h⟩, _, _⟩,
+  { rintro x₁ x₂ (t : dite _ _ _ = dite _ _ _),
+    split_ifs at t with h₁ h₂ h₃ h₄,
+    { subst h₁,
+      subst h₂ },
+    { subst h₁,
+      apply ((f' ⟨x₂, h₂⟩).prop t.symm).elim },
+    { subst h₃,
+      apply ((f' ⟨x₁, h₁⟩).prop t).elim },
+    { injection hfinj (subtype.coe_injective t) } },
+  { intro x,
+    split_ifs,
+    { subst h,
+      simpa [image_rel] using hb },
+    { apply hfr ⟨x, h⟩ } }
+end
+
+/-- Second case of the inductive step: assuming that
+`∃ (A : finset α), A ≠ univ → A.card = (image_rel r A).card`
+and that the statement of Hall's Marriage Theorem
+is true for all `α'` of cardinality ≤ `n`, then it is true for `α`.
+-/
+lemma hall_hard_inductive_step_B [nontrivial α] {n : ℕ} (hn : fintype.card α ≤ n.succ)
+  (hr : ∀ (A : finset α), A.card ≤ (image_rel r A).card)
+  (ha : ∃ (A : finset α), A.nonempty ∧ A ≠ univ ∧ A.card = (image_rel r A).card)
+  (ih : ∀ {α' β' : Type u} [fintype α'] [fintype β'] (r' : α' → β' → Prop)
+    [∀ a, decidable_pred (r' a)], by exactI fintype.card α' ≤ n →
+    by exactI (∀ (A : finset α'), A.card ≤ (image_rel r' A).card) →
+    ∃ (f : α' → β'), function.injective f ∧ ∀ x, r' x (f x)) :
+  ∃ (f : α → β), function.injective f ∧ ∀ x, r x (f x) :=
+begin
+  rcases ha with ⟨A, hA, hnA, huA⟩,
+  let α' := {a' : α // a' ∈ A},
+  let β' := {b' : β // b' ∈ image_rel r A},
+  let r' : α' → β' → Prop := λ a' b', r a' b',
+  have h3 : fintype.card α' ≤ n,
+  { rw [fintype.card_of_subtype A (λ x, iff.rfl), ← nat.lt_succ_iff],
+    have : A ⊂ univ := ⟨subset_univ _, λ t, hnA (le_antisymm (subset_univ _) t)⟩,
+    apply lt_of_lt_of_le (card_lt_card this) hn },
+  have h4 : (∀ (A' : finset α'), A'.card ≤ (image_rel r' A').card),
+  { intro A',
+    have h₁ := hr (A'.image subtype.val),
+    have h₂ : image_rel r (image subtype.val A') ⊆ image subtype.val (image_rel r' A'),
+    { intro t,
+      simp only [image_rel, mem_filter, mem_image, exists_prop, mem_univ, true_and, subtype.exists,
+        and_imp, exists_eq_right, exists_imp_distrib, exists_and_distrib_right],
+      intros a hA hA' rat,
+      refine ⟨_, _, _, hA', rat⟩,
+      simp only [image_rel, mem_filter, mem_univ, true_and, exists_prop],
+      refine ⟨_, hA, rat⟩ },
+    rw card_image_of_injective _ subtype.val_injective at h₁,
+    apply h₁.trans ((card_le_of_subset h₂).trans _),
+    rw card_image_of_injective _ subtype.val_injective },
+  have h' := ih r' h3 h4,
+  rcases h' with ⟨f', hf', hAf'⟩,
+  let α'' := {a'' : α // a'' ∉ A},
+  let β'' := {b'' : β // b'' ∉ image_rel r A},
+  let r'' : α'' → β'' → Prop := λ a'' b'', r a'' b'',
+  have h5 : fintype.card α'' ≤ n,
+  { have : ¬univ ⊆ Aᶜ,
+    { intro t,
+      rcases hA with ⟨a, ha⟩,
+      simpa [ha] using t (mem_univ a) },
+    have : Aᶜ ⊂ univ := ⟨subset_univ _, this⟩,
```

Comment:
Maybe there should be a lemma saying ssubset univ iff there's an element missing to make proofs like this easier